// devexpress-reporting-angular/esm5/lib/dx-reporting-component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Renderer2, ElementRef, ViewChild, Input, Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import * as ko from "knockout";
/**
 * @abstract
 * @template T
 */
var DxReportingComponent = /** @class */ (function () {
    function DxReportingComponent(renderer, platformId) {
        this.renderer = renderer;
        this.platformId = platformId;
        this.height = '800px';
        this.width = '100%';
        this.cssClass = '';
    }
    /**
     * @return {?}
     */
    DxReportingComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (isPlatformBrowser(this.platformId)) {
            this.callbacks = this.getCallbacks();
            /** @type {?} */
            var options = this.getControlOptions();
            /** @type {?} */
            var binding = this.getKoBinding(options);
            this.bindingSender = binding.sender;
            binding.applyBindings(this.control.nativeElement);
        }
    };
    /**
     * @return {?}
     */
    DxReportingComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        ko.cleanNode(this.control.nativeElement);
    };
    /**
     * @return {?}
     */
    DxReportingComponent.prototype.getEventRaiser = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var component = this;
        return function (eventName, args) { return component.callbacks && component.callbacks[eventName] && component.callbacks[eventName].emit({ sender: component.bindingSender, args: args, component: component }); };
    };
    /** @nocollapse */
    DxReportingComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    DxReportingComponent.propDecorators = {
        control: [{ type: ViewChild, args: ["control",] }],
        height: [{ type: Input }],
        width: [{ type: Input }],
        cssClass: [{ type: Input }],
        rtl: [{ type: Input }]
    };
    return DxReportingComponent;
}());
export { DxReportingComponent };
if (false) {
    /** @type {?} */
    DxReportingComponent.prototype.bindingSender;
    /** @type {?} */
    DxReportingComponent.prototype.callbacks;
    /** @type {?} */
    DxReportingComponent.prototype.control;
    /** @type {?} */
    DxReportingComponent.prototype.height;
    /** @type {?} */
    DxReportingComponent.prototype.width;
    /** @type {?} */
    DxReportingComponent.prototype.cssClass;
    /** @type {?} */
    DxReportingComponent.prototype.rtl;
    /** @type {?} */
    DxReportingComponent.prototype.renderer;
    /** @type {?} */
    DxReportingComponent.prototype.platformId;
    /**
     * @abstract
     * @return {?}
     */
    DxReportingComponent.prototype.getControlOptions = function () { };
    /**
     * @abstract
     * @return {?}
     */
    DxReportingComponent.prototype.getCallbacks = function () { };
    /**
     * @abstract
     * @param {?} options
     * @return {?}
     */
    DxReportingComponent.prototype.getKoBinding = function (options) { };
}
