// devexpress-reporting-angular/esm5/lib/dx-report-viewer/dx-report-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Renderer2, ContentChild, Inject, PLATFORM_ID, ViewEncapsulation } from '@angular/core';
import * as ko from 'knockout';
import { Report } from "devexpress-reporting/dx-webdocumentviewer";
import { DxReportingComponent } from '../dx-reporting-component';
import { DxReportViewerRequestOptionsComponent } from './dxrv-request-options';
import { DxReportViewerCallbacksComponent } from './dxrv-callbacks';
import { DxReportViewerMobileModeSettingsComponent } from './dxrv-mobile-mode-settings.component';
import { DxReportViewerRemoteSettingsComponent } from './dxrv-remote-settings.component';
import { DxReportViewerTabPanelSettingsComponent } from './dxrv-tabpanel-settings.component';
var DxReportViewerComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DxReportViewerComponent, _super);
    function DxReportViewerComponent(renderer, platformId) {
        var _this = _super.call(this, renderer, platformId) || this;
        _this.koReportUrl = ko.observable();
        return _this;
    }
    /**
     * @param {?} options
     * @return {?}
     */
    DxReportViewerComponent.prototype.getKoBinding = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return new Report.Preview.JSReportViewerBinding(options, this.getEventRaiser());
    };
    /**
     * @return {?}
     */
    DxReportViewerComponent.prototype.getCallbacks = /**
     * @return {?}
     */
    function () {
        return this.callbacksComponent;
    };
    /**
     * @return {?}
     */
    DxReportViewerComponent.prototype.getControlOptions = /**
     * @return {?}
     */
    function () {
        return {
            rtl: this.rtl,
            isMobile: this.isMobile,
            reportUrl: this.koReportUrl,
            requestOptions: this.requestOptionsComponent,
            mobileModeSettings: this.mobileModeSettingsComponent,
            remoteSettings: this.remoteSettingsComponent,
            tabPanelSettings: this.tabPanelSettingsComponent,
            callbacks: {}
        };
    };
    /**
     * @return {?}
     */
    DxReportViewerComponent.prototype.GetPreviewModel = /**
     * @return {?}
     */
    function () {
        return this.bindingSender.GetPreviewModel();
    };
    Object.defineProperty(DxReportViewerComponent.prototype, "reportUrl", {
        get: /**
         * @return {?}
         */
        function () {
            return this.koReportUrl();
        },
        set: /**
         * @param {?} reportUrl
         * @return {?}
         */
        function (reportUrl) {
            this.koReportUrl(reportUrl);
        },
        enumerable: true,
        configurable: true
    });
    DxReportViewerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dx-report-viewer',
                    encapsulation: ViewEncapsulation.None,
                    template: "<div #control [ngClass]=\"cssClass\" [ngStyle]=\"{'width': width, 'height': height }\">\n  <div data-bind=\"dxReportViewer: $data\"></div>\n</div>",
                    styles: []
                },] },
    ];
    /** @nocollapse */
    DxReportViewerComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    DxReportViewerComponent.propDecorators = {
        requestOptionsComponent: [{ type: ContentChild, args: [DxReportViewerRequestOptionsComponent,] }],
        callbacksComponent: [{ type: ContentChild, args: [DxReportViewerCallbacksComponent,] }],
        mobileModeSettingsComponent: [{ type: ContentChild, args: [DxReportViewerMobileModeSettingsComponent,] }],
        remoteSettingsComponent: [{ type: ContentChild, args: [DxReportViewerRemoteSettingsComponent,] }],
        tabPanelSettingsComponent: [{ type: ContentChild, args: [DxReportViewerTabPanelSettingsComponent,] }],
        isMobile: [{ type: Input }],
        reportUrl: [{ type: Input }]
    };
    return DxReportViewerComponent;
}(DxReportingComponent));
export { DxReportViewerComponent };
if (false) {
    /** @type {?} */
    DxReportViewerComponent.prototype.koReportUrl;
    /** @type {?} */
    DxReportViewerComponent.prototype.requestOptionsComponent;
    /** @type {?} */
    DxReportViewerComponent.prototype.callbacksComponent;
    /** @type {?} */
    DxReportViewerComponent.prototype.mobileModeSettingsComponent;
    /** @type {?} */
    DxReportViewerComponent.prototype.remoteSettingsComponent;
    /** @type {?} */
    DxReportViewerComponent.prototype.tabPanelSettingsComponent;
    /** @type {?} */
    DxReportViewerComponent.prototype.isMobile;
}
