// devexpress-reporting-angular/esm5/lib/dx-report-viewer/dxrv-callbacks.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter } from '@angular/core';
var DxReportViewerCallbacksComponent = /** @class */ (function () {
    function DxReportViewerCallbacksComponent() {
        this.CustomizeElements = new EventEmitter();
        this.PreviewClick = new EventEmitter();
        this.EditingFieldChanged = new EventEmitter();
        this.DocumentReady = new EventEmitter();
        this.CustomizeExportOptions = new EventEmitter();
        this.CustomizeParameterEditors = new EventEmitter();
        this.CustomizeParameterLookUpSource = new EventEmitter();
        this.ParametersReset = new EventEmitter();
        this.ParametersSubmitted = new EventEmitter();
        this.CustomizeMenuActions = new EventEmitter();
        this.CustomizeLocalization = new EventEmitter();
        this.BeforeRender = new EventEmitter();
        this.OnServerError = new EventEmitter();
    }
    DxReportViewerCallbacksComponent.decorators = [
        { type: Component, args: [{
                    template: '',
                    selector: 'dxrv-callbacks'
                },] },
    ];
    DxReportViewerCallbacksComponent.propDecorators = {
        CustomizeElements: [{ type: Output }],
        PreviewClick: [{ type: Output }],
        EditingFieldChanged: [{ type: Output }],
        DocumentReady: [{ type: Output }],
        CustomizeExportOptions: [{ type: Output }],
        CustomizeParameterEditors: [{ type: Output }],
        CustomizeParameterLookUpSource: [{ type: Output }],
        ParametersReset: [{ type: Output }],
        ParametersSubmitted: [{ type: Output }],
        CustomizeMenuActions: [{ type: Output }],
        CustomizeLocalization: [{ type: Output }],
        BeforeRender: [{ type: Output }],
        OnServerError: [{ type: Output }]
    };
    return DxReportViewerCallbacksComponent;
}());
export { DxReportViewerCallbacksComponent };
if (false) {
    /** @type {?} */
    DxReportViewerCallbacksComponent.prototype.CustomizeElements;
    /** @type {?} */
    DxReportViewerCallbacksComponent.prototype.PreviewClick;
    /** @type {?} */
    DxReportViewerCallbacksComponent.prototype.EditingFieldChanged;
    /** @type {?} */
    DxReportViewerCallbacksComponent.prototype.DocumentReady;
    /** @type {?} */
    DxReportViewerCallbacksComponent.prototype.CustomizeExportOptions;
    /** @type {?} */
    DxReportViewerCallbacksComponent.prototype.CustomizeParameterEditors;
    /** @type {?} */
    DxReportViewerCallbacksComponent.prototype.CustomizeParameterLookUpSource;
    /** @type {?} */
    DxReportViewerCallbacksComponent.prototype.ParametersReset;
    /** @type {?} */
    DxReportViewerCallbacksComponent.prototype.ParametersSubmitted;
    /** @type {?} */
    DxReportViewerCallbacksComponent.prototype.CustomizeMenuActions;
    /** @type {?} */
    DxReportViewerCallbacksComponent.prototype.CustomizeLocalization;
    /** @type {?} */
    DxReportViewerCallbacksComponent.prototype.BeforeRender;
    /** @type {?} */
    DxReportViewerCallbacksComponent.prototype.OnServerError;
}
