// devexpress-reporting-angular/esm5/lib/dx-report-viewer/dxrv-request-options.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var DxReportViewerRequestOptionsComponent = /** @class */ (function () {
    function DxReportViewerRequestOptionsComponent() {
        this.invokeAction = "/DXXRDV";
    }
    DxReportViewerRequestOptionsComponent.decorators = [
        { type: Component, args: [{
                    template: '',
                    selector: 'dxrv-request-options'
                },] },
    ];
    DxReportViewerRequestOptionsComponent.propDecorators = {
        invokeAction: [{ type: Input }],
        getLocalizationAction: [{ type: Input }],
        host: [{ type: Input }]
    };
    return DxReportViewerRequestOptionsComponent;
}());
export { DxReportViewerRequestOptionsComponent };
if (false) {
    /** @type {?} */
    DxReportViewerRequestOptionsComponent.prototype.invokeAction;
    /** @type {?} */
    DxReportViewerRequestOptionsComponent.prototype.getLocalizationAction;
    /** @type {?} */
    DxReportViewerRequestOptionsComponent.prototype.host;
}
