// devexpress-reporting-angular/esm5/lib/dx-report-designer/dxrd-request-options.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var DxReportDesignerRequestOptionsComponent = /** @class */ (function () {
    function DxReportDesignerRequestOptionsComponent() {
    }
    DxReportDesignerRequestOptionsComponent.decorators = [
        { type: Component, args: [{
                    template: '',
                    selector: 'dxrd-request-options'
                },] },
    ];
    DxReportDesignerRequestOptionsComponent.propDecorators = {
        getDesignerModelAction: [{ type: Input }],
        getLocalizationAction: [{ type: Input }],
        host: [{ type: Input }]
    };
    return DxReportDesignerRequestOptionsComponent;
}());
export { DxReportDesignerRequestOptionsComponent };
if (false) {
    /** @type {?} */
    DxReportDesignerRequestOptionsComponent.prototype.getDesignerModelAction;
    /** @type {?} */
    DxReportDesignerRequestOptionsComponent.prototype.getLocalizationAction;
    /** @type {?} */
    DxReportDesignerRequestOptionsComponent.prototype.host;
}
