// devexpress-reporting-angular/esm2015/lib/dx-report-designer/dx-report-designer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Renderer2, ContentChild, Inject, PLATFORM_ID, ViewEncapsulation } from '@angular/core';
import * as ko from 'knockout';
import { Designer } from "devexpress-reporting/dx-reportdesigner";
import { DxReportingComponent } from '../dx-reporting-component';
import { DxReportDesignerRequestOptionsComponent } from './dxrd-request-options';
import { DxReportDesignerCallbacksComponent } from './dxrd-callbacks';
export class DxReportDesignerComponent extends DxReportingComponent {
    /**
     * @param {?} renderer
     * @param {?} platformId
     */
    constructor(renderer, platformId) {
        super(renderer, platformId);
        this.koReportUrl = ko.observable();
    }
    /**
     * @return {?}
     */
    getCallbacks() {
        return this.callbacksComponent;
    }
    /**
     * @param {?} options
     * @return {?}
     */
    getKoBinding(options) {
        return new Designer.Report.JSReportDesignerBinding(options, this.getEventRaiser());
    }
    /**
     * @return {?}
     */
    getControlOptions() {
        return {
            allowMDI: this.allowMDI,
            reportUrl: this.koReportUrl,
            requestOptions: this.requestOptions,
            callbacks: { designer: {}, preview: {} }
        };
    }
    /**
     * @return {?}
     */
    GetDesignerModel() {
        return this.bindingSender.GetDesignerModel();
    }
    /**
     * @return {?}
     */
    GetPreviewModel() {
        return this.bindingSender.GetPreviewModel();
    }
    /**
     * @return {?}
     */
    get reportUrl() {
        return this.koReportUrl();
    }
    /**
     * @param {?} reportUrl
     * @return {?}
     */
    set reportUrl(reportUrl) {
        this.koReportUrl(reportUrl);
    }
}
DxReportDesignerComponent.decorators = [
    { type: Component, args: [{
                selector: 'dx-report-designer',
                encapsulation: ViewEncapsulation.None,
                template: `<div #control [ngClass]="cssClass" [ngStyle]="{'width': width, 'height': height }">
  <div data-bind="dxReportDesigner: $data"></div>
</div>`,
                styles: []
            },] },
];
/** @nocollapse */
DxReportDesignerComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
DxReportDesignerComponent.propDecorators = {
    requestOptions: [{ type: ContentChild, args: [DxReportDesignerRequestOptionsComponent,] }],
    callbacksComponent: [{ type: ContentChild, args: [DxReportDesignerCallbacksComponent,] }],
    allowMDI: [{ type: Input }],
    reportUrl: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DxReportDesignerComponent.prototype.koReportUrl;
    /** @type {?} */
    DxReportDesignerComponent.prototype.requestOptions;
    /** @type {?} */
    DxReportDesignerComponent.prototype.callbacksComponent;
    /** @type {?} */
    DxReportDesignerComponent.prototype.allowMDI;
}
