// devexpress-reporting-angular/esm2015/lib/dx-report-designer/dxrd-callbacks.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter } from '@angular/core';
export class DxReportDesignerCallbacksComponent {
    constructor() {
        this.CustomizeElements = new EventEmitter();
        this.ExitDesigner = new EventEmitter();
        this.ReportSaving = new EventEmitter();
        this.ReportSaved = new EventEmitter();
        this.ReportOpened = new EventEmitter();
        this.ReportOpening = new EventEmitter();
        this.TabChanged = new EventEmitter();
        this.ComponentAdded = new EventEmitter();
        this.CustomizeParameterEditors = new EventEmitter();
        this.CustomizeSaveDialog = new EventEmitter();
        this.CustomizeOpenDialog = new EventEmitter();
        this.CustomizeToolbox = new EventEmitter();
        this.CustomizeFieldListActions = new EventEmitter();
        this.CustomizeMenuActions = new EventEmitter();
        this.CustomizeLocalization = new EventEmitter();
        this.BeforeRender = new EventEmitter();
        this.OnServerError = new EventEmitter();
        this.PreviewClick = new EventEmitter();
        this.PreviewCustomizeElements = new EventEmitter();
        this.PreviewCustomizeMenuActions = new EventEmitter();
        this.PreviewEditingFieldChanged = new EventEmitter();
        this.PreviewDocumentReady = new EventEmitter();
        this.PreviewParametersReset = new EventEmitter();
        this.PreviewCustomizeExportOptions = new EventEmitter();
        this.ParametersSubmitted = new EventEmitter();
    }
}
DxReportDesignerCallbacksComponent.decorators = [
    { type: Component, args: [{
                template: '',
                selector: 'dxrd-callbacks'
            },] },
];
DxReportDesignerCallbacksComponent.propDecorators = {
    CustomizeElements: [{ type: Output }],
    ExitDesigner: [{ type: Output }],
    ReportSaving: [{ type: Output }],
    ReportSaved: [{ type: Output }],
    ReportOpened: [{ type: Output }],
    ReportOpening: [{ type: Output }],
    TabChanged: [{ type: Output }],
    ComponentAdded: [{ type: Output }],
    CustomizeParameterEditors: [{ type: Output }],
    CustomizeSaveDialog: [{ type: Output }],
    CustomizeOpenDialog: [{ type: Output }],
    CustomizeToolbox: [{ type: Output }],
    CustomizeFieldListActions: [{ type: Output }],
    CustomizeMenuActions: [{ type: Output }],
    CustomizeLocalization: [{ type: Output }],
    BeforeRender: [{ type: Output }],
    OnServerError: [{ type: Output }],
    PreviewClick: [{ type: Output }],
    PreviewCustomizeElements: [{ type: Output }],
    PreviewCustomizeMenuActions: [{ type: Output }],
    PreviewEditingFieldChanged: [{ type: Output }],
    PreviewDocumentReady: [{ type: Output }],
    PreviewParametersReset: [{ type: Output }],
    PreviewCustomizeExportOptions: [{ type: Output }],
    ParametersSubmitted: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.CustomizeElements;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.ExitDesigner;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.ReportSaving;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.ReportSaved;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.ReportOpened;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.ReportOpening;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.TabChanged;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.ComponentAdded;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.CustomizeParameterEditors;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.CustomizeSaveDialog;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.CustomizeOpenDialog;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.CustomizeToolbox;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.CustomizeFieldListActions;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.CustomizeMenuActions;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.CustomizeLocalization;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.BeforeRender;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.OnServerError;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.PreviewClick;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.PreviewCustomizeElements;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.PreviewCustomizeMenuActions;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.PreviewEditingFieldChanged;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.PreviewDocumentReady;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.PreviewParametersReset;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.PreviewCustomizeExportOptions;
    /** @type {?} */
    DxReportDesignerCallbacksComponent.prototype.ParametersSubmitted;
}
