// devexpress-reporting-angular/esm5/devexpress-reporting-angular.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { DxReportViewerModule, DxReportDesignerModule, DxReportDesignerComponent, DxReportViewerComponent } from './public_api';
export { DxReportDesignerCallbacksComponent as ɵg } from './lib/dx-report-designer/dxrd-callbacks';
export { DxReportDesignerRequestOptionsComponent as ɵh } from './lib/dx-report-designer/dxrd-request-options';
export { DxReportViewerCallbacksComponent as ɵc } from './lib/dx-report-viewer/dxrv-callbacks';
export { DxReportViewerMobileModeSettingsComponent as ɵd } from './lib/dx-report-viewer/dxrv-mobile-mode-settings.component';
export { DxReportViewerRemoteSettingsComponent as ɵe } from './lib/dx-report-viewer/dxrv-remote-settings.component';
export { DxReportViewerRequestOptionsComponent as ɵb } from './lib/dx-report-viewer/dxrv-request-options';
export { DxReportViewerTabPanelSettingsComponent as ɵf } from './lib/dx-report-viewer/dxrv-tabpanel-settings.component';
export { DxReportingComponent as ɵa } from './lib/dx-reporting-component';
