// devexpress-reporting-angular/fesm5/devexpress-reporting-angular.js
import { Component, Output, EventEmitter, Input, Renderer2, ViewChild, Inject, PLATFORM_ID, ContentChild, ViewEncapsulation, NgModule } from '@angular/core';
import { isPlatformBrowser, CommonModule } from '@angular/common';
import { cleanNode, observable } from 'knockout';
import { __extends } from 'tslib';
import { Report } from 'devexpress-reporting/dx-webdocumentviewer';
import { Designer } from 'devexpress-reporting/dx-reportdesigner';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var DxReportViewerCallbacksComponent = /** @class */ (function () {
    function DxReportViewerCallbacksComponent() {
        this.CustomizeElements = new EventEmitter();
        this.PreviewClick = new EventEmitter();
        this.EditingFieldChanged = new EventEmitter();
        this.DocumentReady = new EventEmitter();
        this.CustomizeExportOptions = new EventEmitter();
        this.CustomizeParameterEditors = new EventEmitter();
        this.CustomizeParameterLookUpSource = new EventEmitter();
        this.ParametersReset = new EventEmitter();
        this.ParametersSubmitted = new EventEmitter();
        this.CustomizeMenuActions = new EventEmitter();
        this.CustomizeLocalization = new EventEmitter();
        this.BeforeRender = new EventEmitter();
        this.OnServerError = new EventEmitter();
    }
    DxReportViewerCallbacksComponent.decorators = [
        { type: Component, args: [{
                    template: '',
                    selector: 'dxrv-callbacks'
                },] },
    ];
    DxReportViewerCallbacksComponent.propDecorators = {
        CustomizeElements: [{ type: Output }],
        PreviewClick: [{ type: Output }],
        EditingFieldChanged: [{ type: Output }],
        DocumentReady: [{ type: Output }],
        CustomizeExportOptions: [{ type: Output }],
        CustomizeParameterEditors: [{ type: Output }],
        CustomizeParameterLookUpSource: [{ type: Output }],
        ParametersReset: [{ type: Output }],
        ParametersSubmitted: [{ type: Output }],
        CustomizeMenuActions: [{ type: Output }],
        CustomizeLocalization: [{ type: Output }],
        BeforeRender: [{ type: Output }],
        OnServerError: [{ type: Output }]
    };
    return DxReportViewerCallbacksComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var DxReportViewerRequestOptionsComponent = /** @class */ (function () {
    function DxReportViewerRequestOptionsComponent() {
        this.invokeAction = "/DXXRDV";
    }
    DxReportViewerRequestOptionsComponent.decorators = [
        { type: Component, args: [{
                    template: '',
                    selector: 'dxrv-request-options'
                },] },
    ];
    DxReportViewerRequestOptionsComponent.propDecorators = {
        invokeAction: [{ type: Input }],
        getLocalizationAction: [{ type: Input }],
        host: [{ type: Input }]
    };
    return DxReportViewerRequestOptionsComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @abstract
 * @template T
 */
var DxReportingComponent = /** @class */ (function () {
    function DxReportingComponent(renderer, platformId) {
        this.renderer = renderer;
        this.platformId = platformId;
        this.height = '800px';
        this.width = '100%';
        this.cssClass = '';
    }
    /**
     * @return {?}
     */
    DxReportingComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (isPlatformBrowser(this.platformId)) {
            this.callbacks = this.getCallbacks();
            /** @type {?} */
            var options = this.getControlOptions();
            /** @type {?} */
            var binding = this.getKoBinding(options);
            this.bindingSender = binding.sender;
            binding.applyBindings(this.control.nativeElement);
        }
    };
    /**
     * @return {?}
     */
    DxReportingComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        cleanNode(this.control.nativeElement);
    };
    /**
     * @return {?}
     */
    DxReportingComponent.prototype.getEventRaiser = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var component = this;
        return function (eventName, args) { return component.callbacks && component.callbacks[eventName] && component.callbacks[eventName].emit({ sender: component.bindingSender, args: args, component: component }); };
    };
    /** @nocollapse */
    DxReportingComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    DxReportingComponent.propDecorators = {
        control: [{ type: ViewChild, args: ["control",] }],
        height: [{ type: Input }],
        width: [{ type: Input }],
        cssClass: [{ type: Input }],
        rtl: [{ type: Input }]
    };
    return DxReportingComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var DxReportViewerMobileModeSettingsComponent = /** @class */ (function () {
    function DxReportViewerMobileModeSettingsComponent() {
    }
    DxReportViewerMobileModeSettingsComponent.decorators = [
        { type: Component, args: [{
                    template: '',
                    selector: 'dxrv-mobile-mode-settings'
                },] },
    ];
    DxReportViewerMobileModeSettingsComponent.propDecorators = {
        readerMode: [{ type: Input }],
        animationEnabled: [{ type: Input }]
    };
    return DxReportViewerMobileModeSettingsComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var DxReportViewerRemoteSettingsComponent = /** @class */ (function () {
    function DxReportViewerRemoteSettingsComponent() {
    }
    DxReportViewerRemoteSettingsComponent.decorators = [
        { type: Component, args: [{
                    template: '',
                    selector: 'dxrv-remote-settings'
                },] },
    ];
    DxReportViewerRemoteSettingsComponent.propDecorators = {
        authToken: [{ type: Input }],
        serverUri: [{ type: Input }]
    };
    return DxReportViewerRemoteSettingsComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var DxReportViewerTabPanelSettingsComponent = /** @class */ (function () {
    function DxReportViewerTabPanelSettingsComponent() {
    }
    DxReportViewerTabPanelSettingsComponent.decorators = [
        { type: Component, args: [{
                    template: '',
                    selector: 'dxrv-tabpanel-settings'
                },] },
    ];
    DxReportViewerTabPanelSettingsComponent.propDecorators = {
        position: [{ type: Input }],
        width: [{ type: Input }]
    };
    return DxReportViewerTabPanelSettingsComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var DxReportViewerComponent = /** @class */ (function (_super) {
    __extends(DxReportViewerComponent, _super);
    function DxReportViewerComponent(renderer, platformId) {
        var _this = _super.call(this, renderer, platformId) || this;
        _this.koReportUrl = observable();
        return _this;
    }
    /**
     * @param {?} options
     * @return {?}
     */
    DxReportViewerComponent.prototype.getKoBinding = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return new Report.Preview.JSReportViewerBinding(options, this.getEventRaiser());
    };
    /**
     * @return {?}
     */
    DxReportViewerComponent.prototype.getCallbacks = /**
     * @return {?}
     */
    function () {
        return this.callbacksComponent;
    };
    /**
     * @return {?}
     */
    DxReportViewerComponent.prototype.getControlOptions = /**
     * @return {?}
     */
    function () {
        return {
            rtl: this.rtl,
            isMobile: this.isMobile,
            reportUrl: this.koReportUrl,
            requestOptions: this.requestOptionsComponent,
            mobileModeSettings: this.mobileModeSettingsComponent,
            remoteSettings: this.remoteSettingsComponent,
            tabPanelSettings: this.tabPanelSettingsComponent,
            callbacks: {}
        };
    };
    /**
     * @return {?}
     */
    DxReportViewerComponent.prototype.GetPreviewModel = /**
     * @return {?}
     */
    function () {
        return this.bindingSender.GetPreviewModel();
    };
    Object.defineProperty(DxReportViewerComponent.prototype, "reportUrl", {
        get: /**
         * @return {?}
         */
        function () {
            return this.koReportUrl();
        },
        set: /**
         * @param {?} reportUrl
         * @return {?}
         */
        function (reportUrl) {
            this.koReportUrl(reportUrl);
        },
        enumerable: true,
        configurable: true
    });
    DxReportViewerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dx-report-viewer',
                    encapsulation: ViewEncapsulation.None,
                    template: "<div #control [ngClass]=\"cssClass\" [ngStyle]=\"{'width': width, 'height': height }\">\n  <div data-bind=\"dxReportViewer: $data\"></div>\n</div>",
                    styles: []
                },] },
    ];
    /** @nocollapse */
    DxReportViewerComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    DxReportViewerComponent.propDecorators = {
        requestOptionsComponent: [{ type: ContentChild, args: [DxReportViewerRequestOptionsComponent,] }],
        callbacksComponent: [{ type: ContentChild, args: [DxReportViewerCallbacksComponent,] }],
        mobileModeSettingsComponent: [{ type: ContentChild, args: [DxReportViewerMobileModeSettingsComponent,] }],
        remoteSettingsComponent: [{ type: ContentChild, args: [DxReportViewerRemoteSettingsComponent,] }],
        tabPanelSettingsComponent: [{ type: ContentChild, args: [DxReportViewerTabPanelSettingsComponent,] }],
        isMobile: [{ type: Input }],
        reportUrl: [{ type: Input }]
    };
    return DxReportViewerComponent;
}(DxReportingComponent));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var DxReportViewerModule = /** @class */ (function () {
    function DxReportViewerModule() {
    }
    DxReportViewerModule.decorators = [
        { type: NgModule, args: [{
                    id: "DxReportViewerModule",
                    imports: [CommonModule],
                    declarations: [
                        DxReportViewerComponent,
                        DxReportViewerCallbacksComponent,
                        DxReportViewerMobileModeSettingsComponent,
                        DxReportViewerRemoteSettingsComponent,
                        DxReportViewerRequestOptionsComponent,
                        DxReportViewerTabPanelSettingsComponent
                    ],
                    exports: [
                        DxReportViewerComponent,
                        DxReportViewerCallbacksComponent,
                        DxReportViewerMobileModeSettingsComponent,
                        DxReportViewerRemoteSettingsComponent,
                        DxReportViewerRequestOptionsComponent,
                        DxReportViewerTabPanelSettingsComponent
                    ]
                },] },
    ];
    return DxReportViewerModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var DxReportDesignerRequestOptionsComponent = /** @class */ (function () {
    function DxReportDesignerRequestOptionsComponent() {
    }
    DxReportDesignerRequestOptionsComponent.decorators = [
        { type: Component, args: [{
                    template: '',
                    selector: 'dxrd-request-options'
                },] },
    ];
    DxReportDesignerRequestOptionsComponent.propDecorators = {
        getDesignerModelAction: [{ type: Input }],
        getLocalizationAction: [{ type: Input }],
        host: [{ type: Input }]
    };
    return DxReportDesignerRequestOptionsComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var DxReportDesignerCallbacksComponent = /** @class */ (function () {
    function DxReportDesignerCallbacksComponent() {
        this.CustomizeElements = new EventEmitter();
        this.ExitDesigner = new EventEmitter();
        this.ReportSaving = new EventEmitter();
        this.ReportSaved = new EventEmitter();
        this.ReportOpened = new EventEmitter();
        this.ReportOpening = new EventEmitter();
        this.TabChanged = new EventEmitter();
        this.ComponentAdded = new EventEmitter();
        this.CustomizeParameterEditors = new EventEmitter();
        this.CustomizeSaveDialog = new EventEmitter();
        this.CustomizeOpenDialog = new EventEmitter();
        this.CustomizeToolbox = new EventEmitter();
        this.CustomizeFieldListActions = new EventEmitter();
        this.CustomizeMenuActions = new EventEmitter();
        this.CustomizeLocalization = new EventEmitter();
        this.BeforeRender = new EventEmitter();
        this.OnServerError = new EventEmitter();
        this.PreviewClick = new EventEmitter();
        this.PreviewCustomizeElements = new EventEmitter();
        this.PreviewCustomizeMenuActions = new EventEmitter();
        this.PreviewEditingFieldChanged = new EventEmitter();
        this.PreviewDocumentReady = new EventEmitter();
        this.PreviewParametersReset = new EventEmitter();
        this.PreviewCustomizeExportOptions = new EventEmitter();
        this.ParametersSubmitted = new EventEmitter();
    }
    DxReportDesignerCallbacksComponent.decorators = [
        { type: Component, args: [{
                    template: '',
                    selector: 'dxrd-callbacks'
                },] },
    ];
    DxReportDesignerCallbacksComponent.propDecorators = {
        CustomizeElements: [{ type: Output }],
        ExitDesigner: [{ type: Output }],
        ReportSaving: [{ type: Output }],
        ReportSaved: [{ type: Output }],
        ReportOpened: [{ type: Output }],
        ReportOpening: [{ type: Output }],
        TabChanged: [{ type: Output }],
        ComponentAdded: [{ type: Output }],
        CustomizeParameterEditors: [{ type: Output }],
        CustomizeSaveDialog: [{ type: Output }],
        CustomizeOpenDialog: [{ type: Output }],
        CustomizeToolbox: [{ type: Output }],
        CustomizeFieldListActions: [{ type: Output }],
        CustomizeMenuActions: [{ type: Output }],
        CustomizeLocalization: [{ type: Output }],
        BeforeRender: [{ type: Output }],
        OnServerError: [{ type: Output }],
        PreviewClick: [{ type: Output }],
        PreviewCustomizeElements: [{ type: Output }],
        PreviewCustomizeMenuActions: [{ type: Output }],
        PreviewEditingFieldChanged: [{ type: Output }],
        PreviewDocumentReady: [{ type: Output }],
        PreviewParametersReset: [{ type: Output }],
        PreviewCustomizeExportOptions: [{ type: Output }],
        ParametersSubmitted: [{ type: Output }]
    };
    return DxReportDesignerCallbacksComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var DxReportDesignerComponent = /** @class */ (function (_super) {
    __extends(DxReportDesignerComponent, _super);
    function DxReportDesignerComponent(renderer, platformId) {
        var _this = _super.call(this, renderer, platformId) || this;
        _this.koReportUrl = observable();
        return _this;
    }
    /**
     * @return {?}
     */
    DxReportDesignerComponent.prototype.getCallbacks = /**
     * @return {?}
     */
    function () {
        return this.callbacksComponent;
    };
    /**
     * @param {?} options
     * @return {?}
     */
    DxReportDesignerComponent.prototype.getKoBinding = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return new Designer.Report.JSReportDesignerBinding(options, this.getEventRaiser());
    };
    /**
     * @return {?}
     */
    DxReportDesignerComponent.prototype.getControlOptions = /**
     * @return {?}
     */
    function () {
        return {
            allowMDI: this.allowMDI,
            reportUrl: this.koReportUrl,
            requestOptions: this.requestOptions,
            callbacks: { designer: {}, preview: {} }
        };
    };
    /**
     * @return {?}
     */
    DxReportDesignerComponent.prototype.GetDesignerModel = /**
     * @return {?}
     */
    function () {
        return this.bindingSender.GetDesignerModel();
    };
    /**
     * @return {?}
     */
    DxReportDesignerComponent.prototype.GetPreviewModel = /**
     * @return {?}
     */
    function () {
        return this.bindingSender.GetPreviewModel();
    };
    Object.defineProperty(DxReportDesignerComponent.prototype, "reportUrl", {
        get: /**
         * @return {?}
         */
        function () {
            return this.koReportUrl();
        },
        set: /**
         * @param {?} reportUrl
         * @return {?}
         */
        function (reportUrl) {
            this.koReportUrl(reportUrl);
        },
        enumerable: true,
        configurable: true
    });
    DxReportDesignerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dx-report-designer',
                    encapsulation: ViewEncapsulation.None,
                    template: "<div #control [ngClass]=\"cssClass\" [ngStyle]=\"{'width': width, 'height': height }\">\n  <div data-bind=\"dxReportDesigner: $data\"></div>\n</div>",
                    styles: []
                },] },
    ];
    /** @nocollapse */
    DxReportDesignerComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    DxReportDesignerComponent.propDecorators = {
        requestOptions: [{ type: ContentChild, args: [DxReportDesignerRequestOptionsComponent,] }],
        callbacksComponent: [{ type: ContentChild, args: [DxReportDesignerCallbacksComponent,] }],
        allowMDI: [{ type: Input }],
        reportUrl: [{ type: Input }]
    };
    return DxReportDesignerComponent;
}(DxReportingComponent));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var DxReportDesignerModule = /** @class */ (function () {
    function DxReportDesignerModule() {
    }
    DxReportDesignerModule.decorators = [
        { type: NgModule, args: [{
                    id: "DxReportDesignerModule",
                    imports: [CommonModule],
                    declarations: [
                        DxReportDesignerCallbacksComponent,
                        DxReportDesignerRequestOptionsComponent,
                        DxReportDesignerComponent
                    ],
                    exports: [
                        DxReportDesignerCallbacksComponent,
                        DxReportDesignerRequestOptionsComponent,
                        DxReportDesignerComponent
                    ]
                },] },
    ];
    return DxReportDesignerModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { DxReportViewerModule, DxReportDesignerModule, DxReportDesignerComponent, DxReportViewerComponent, DxReportDesignerCallbacksComponent as ɵg, DxReportDesignerRequestOptionsComponent as ɵh, DxReportViewerCallbacksComponent as ɵc, DxReportViewerMobileModeSettingsComponent as ɵd, DxReportViewerRemoteSettingsComponent as ɵe, DxReportViewerRequestOptionsComponent as ɵb, DxReportViewerTabPanelSettingsComponent as ɵf, DxReportingComponent as ɵa };
