// devexpress-reporting-angular/esm2015/lib/dx-report-viewer/dx-report-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Renderer2, ContentChild, Inject, PLATFORM_ID, ViewEncapsulation } from '@angular/core';
import * as ko from 'knockout';
import { Report } from "devexpress-reporting/dx-webdocumentviewer";
import { DxReportingComponent } from '../dx-reporting-component';
import { DxReportViewerRequestOptionsComponent } from './dxrv-request-options';
import { DxReportViewerCallbacksComponent } from './dxrv-callbacks';
import { DxReportViewerMobileModeSettingsComponent } from './dxrv-mobile-mode-settings.component';
import { DxReportViewerRemoteSettingsComponent } from './dxrv-remote-settings.component';
import { DxReportViewerTabPanelSettingsComponent } from './dxrv-tabpanel-settings.component';
export class DxReportViewerComponent extends DxReportingComponent {
    /**
     * @param {?} renderer
     * @param {?} platformId
     */
    constructor(renderer, platformId) {
        super(renderer, platformId);
        this.koReportUrl = ko.observable();
    }
    /**
     * @param {?} options
     * @return {?}
     */
    getKoBinding(options) {
        return new Report.Preview.JSReportViewerBinding(options, this.getEventRaiser());
    }
    /**
     * @return {?}
     */
    getCallbacks() {
        return this.callbacksComponent;
    }
    /**
     * @return {?}
     */
    getControlOptions() {
        return {
            rtl: this.rtl,
            isMobile: this.isMobile,
            reportUrl: this.koReportUrl,
            requestOptions: this.requestOptionsComponent,
            mobileModeSettings: this.mobileModeSettingsComponent,
            remoteSettings: this.remoteSettingsComponent,
            tabPanelSettings: this.tabPanelSettingsComponent,
            callbacks: {}
        };
    }
    /**
     * @return {?}
     */
    GetPreviewModel() {
        return this.bindingSender.GetPreviewModel();
    }
    /**
     * @return {?}
     */
    get reportUrl() {
        return this.koReportUrl();
    }
    /**
     * @param {?} reportUrl
     * @return {?}
     */
    set reportUrl(reportUrl) {
        this.koReportUrl(reportUrl);
    }
}
DxReportViewerComponent.decorators = [
    { type: Component, args: [{
                selector: 'dx-report-viewer',
                encapsulation: ViewEncapsulation.None,
                template: `<div #control [ngClass]="cssClass" [ngStyle]="{'width': width, 'height': height }">
  <div data-bind="dxReportViewer: $data"></div>
</div>`,
                styles: []
            },] },
];
/** @nocollapse */
DxReportViewerComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
DxReportViewerComponent.propDecorators = {
    requestOptionsComponent: [{ type: ContentChild, args: [DxReportViewerRequestOptionsComponent,] }],
    callbacksComponent: [{ type: ContentChild, args: [DxReportViewerCallbacksComponent,] }],
    mobileModeSettingsComponent: [{ type: ContentChild, args: [DxReportViewerMobileModeSettingsComponent,] }],
    remoteSettingsComponent: [{ type: ContentChild, args: [DxReportViewerRemoteSettingsComponent,] }],
    tabPanelSettingsComponent: [{ type: ContentChild, args: [DxReportViewerTabPanelSettingsComponent,] }],
    isMobile: [{ type: Input }],
    reportUrl: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DxReportViewerComponent.prototype.koReportUrl;
    /** @type {?} */
    DxReportViewerComponent.prototype.requestOptionsComponent;
    /** @type {?} */
    DxReportViewerComponent.prototype.callbacksComponent;
    /** @type {?} */
    DxReportViewerComponent.prototype.mobileModeSettingsComponent;
    /** @type {?} */
    DxReportViewerComponent.prototype.remoteSettingsComponent;
    /** @type {?} */
    DxReportViewerComponent.prototype.tabPanelSettingsComponent;
    /** @type {?} */
    DxReportViewerComponent.prototype.isMobile;
}
